dandi/organize.py

```diff
@@ -96,8 +104,83 @@ def create_unique_filenames_from_metadata(
+                msg_detailed + "\nPlease adjust/provide metadata in your .nwb "
+                "files to disambiguate or rerun allowing adding object_id."
+            )
+        _assign_obj_id(metadata, non_unique)
+        _assign_dandi_names(
+            metadata, mandatory, (mandatory_if_not_empty or []) + ["obj_id"]
+        )
+        non_unique = _get_non_unique_paths(metadata)
+        if non_unique:
+            raise OrganizeImpossibleError(
+                "Even after adding obj_id we ended up with non-unique file names. "
+                "Should have not happened: %s" % str(non_unique)
+            )
+    return metadata
+
+
+def _assign_obj_id(metadata, non_unique):
+    msg = "%d out of %d paths are not unique" % (len(non_unique), len(metadata))
 
+    lgr.info(msg + ". We will try adding _obj- based on crc32 of object_id")
+    seen_obj_ids = {}  # obj_id: object_id
+    seen_object_ids = {}  # object_id: path
+    for r in metadata:
+        if r["dandi_path"] in non_unique:
+            try:
+                object_id = get_object_id(r["path"])
+            except KeyError:
+                raise OrganizeImpossibleError(
+                    msg + ". We tried to use object_id but it is absent in %r. "
+                    "It is either not .nwb file or produced by older *nwb libraries. "
+                    "You must re-save files e.g. using recent pynwb." % (r["path"])
+                )
+
+            if not object_id:
+                raise OrganizeImpossibleError(
+                    msg + ". We tried to use object_id but it was %r for %s. "
+                    " You might need to re-save files using recent pynwb"
+                    % (object_id, r["path"])
+                )
+            # shorter version
+            obj_id = np.base_repr(binascii.crc32(object_id.encode("ascii")), 36).lower()
+            if obj_id in seen_obj_ids:
+                seen_object_id = seen_obj_ids[obj_id]
+                if seen_object_id == object_id:
+                    raise OrganizeImpossibleError(
+                        "Two files (%r and %r) have the same object_id %s. "
+                        "Must not happen. Either files are duplicates "
+                        "(remove one) or were not saved correctly using "
+                        "recent pynwb"
+                        % (r["path"], seen_object_ids[object_id], object_id)
+                    )
+                else:
+                    raise RuntimeError(
+                        f"Wrong assumption by DANDI developers that first "
+                        f"CRC32 checksum of object_id would be sufficient.  Please "
+                        f"report: {obj_id} the same for "
+                        f"{seen_object_ids[seen_object_id]}={seen_object_id} "
+                        f"{r['path']}={object_id} "
+                    )
+            r["obj_id"] = obj_id
+            seen_obj_ids[obj_id] = object_id
+            seen_object_ids[object_id] = r["path"]
```

Comment:
it seems this could benefit from some code improvement. what is the difference between the two variables. semantically seem similar.

Reply:
`obj_id` is the shortened version of `object_id`.  There is a comment by those dictionaries definition saying that first maps from short to original, the 2nd one from original to the path.  Indeed names could be better. E.g. `obj_to_object_ids` and `object_id_paths`